tests/unit/tls_test.cc

```diff
@@ -2097,3 +2097,56 @@ SEASTAR_THREAD_TEST_CASE(test_send_recv_alloc_limits) {
         BOOST_CHECK_EQUAL(stats_after.large_allocations(), stats_before.large_allocations());
     }
 }
+
+static std::pair<connected_socket, connected_socket> tls_socketpair() {
+    auto certs = ::make_shared<tls::server_credentials>(::make_shared<tls::dh_params>());
+    certs->set_x509_key_file(certfile("test.crt"), certfile("test.key"), tls::x509_crt_format::PEM).get();
+
+    ::listen_options opts;
+    opts.reuse_address = true;
+    auto addr = ::make_ipv4_address( {0x7f000001, 4712});
+    auto ss = tls::listen(certs, addr, opts);
+    tls::credentials_builder b;
+    b.set_x509_trust_file(certfile("catest.pem"), tls::x509_crt_format::PEM).get();
+
+    auto cf = tls::connect(b.build_certificate_credentials(), addr);
+    auto ar = ss.accept().get();
+    auto cs = cf.get();
+
+    return std::make_pair(std::move(ar.connection), std::move(cs));
+}
+
+SEASTAR_THREAD_TEST_CASE(test_session_close_with_unread_data) {
+    auto p = tls_socketpair();
+
+    auto c = seastar::async([c = std::move(p.first)] mutable {
+        auto in = c.input();
+        auto b = in.read_exactly(1).get();
+        in.close().get();
+        c.shutdown_output();
+    });
+
+    auto s = seastar::async([s = std::move(p.second)] mutable {
+        auto out = s.output();
+        size_t bytes_sent = 0;
+        auto buf = temporary_buffer<char>(1024);
+        std::memset(buf.get_write(), '\0', buf.size());
+        auto start = std::chrono::steady_clock::now();
+        while (true) {
+            try {
+                out.write(buf.get(), buf.size()).get();
+                out.flush().get();
+                bytes_sent += buf.size();
+            } catch (...) {
+                break;
+            }
+        }
+        auto delay = std::chrono::duration_cast<std::chrono::duration<double>>(std::chrono::steady_clock::now() - start);
+        fmt::print("Wrote {} bytes in {:.3f} seconds\n", bytes_sent, delay.count());
+        out.close().handle_exception([] (auto x) {}).get();
+        s.shutdown_input();
+        BOOST_CHECK_LT(delay.count(), 1.0);
```

Comment:
This seems like a great source of flakey failures...

Reply:
Maybe record a timestamp on the sender side and compare the time elapsed between closer and closee? Though still subject to any weird delay in the network... (though 1s delay even on an overloaded CI machine is unlikely...)

Reply:
Agree about potential flakiness :(
I was inspired by test_skip_wait_for_eof's timeout of 1s to measure similar thing

Reply:
Yeah, that one is not perfect either...

Reply:
No, but you can record when the other side calls "close", and count from there. At least keeps the potential weird delays to the actual wire time from then.